File: Integrations/src/main/java/com/inverce/mod/integrations/support/recycler/DataBinder.java
package com.inverce.mod.integrations.support.recycler;

import android.content.res.Resources;
import android.os.Build;
import android.support.annotation.IdRes;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.inverce.mod.core.IM;
import com.inverce.mod.core.Ui;
import com.inverce.mod.core.functional.IFunction;
import com.inverce.mod.integrations.support.annotations.IBind;
import com.inverce.mod.integrations.support.annotations.MapValue;
import com.inverce.mod.integrations.support.annotations.MapValue.ToDrawable;
import com.inverce.mod.integrations.support.annotations.MapValue.ToDrawableRes;
import com.inverce.mod.integrations.support.annotations.MapValue.ToStringRes;

import java.util.ArrayList;
import java.util.List;

/**
 * Created on 12/10/2017.
 */
public class DataBinder<T> implements IBind<T, BindViewHolder> {
    ToView<String, ImageView> loadImage;

    List<IBind<T, BindViewHolder>> tasks;

    public Resources res() {
        return IM.resources();
    }

    public DataBinder() {
        loadImage = (u, i) -> {
            throw new IllegalStateException("Image processor not specified");
        };
        tasks = new ArrayList<>();
    }

    public <V extends android.view.View> DataBinder<T> bind(ToHolder<T> bind) {
        tasks.add((holder, item, position) -> bind.bind(item, holder));
        return this;
    }

    public <V extends android.view.View> DataBinder<T> bind(IFunction<BindViewHolder, V> view, ToView<T, V> bind) {
        tasks.add((holder, item, position) -> {
            V v = view.apply(holder);
            if (v != null) bind.bind(item, v);
        });
        return this;
    }

    public <V extends android.view.View> DataBinder<T> bind(@IdRes int res, ToView<T, V> bind) {
        return bind(h -> h.get(res), bind);
    }

    public DataBinder<T> bindText(@IdRes int id, MapValue<T, String> map) {
        return bind(p -> (TextView) p.get(id), (item, view) -> view.setText(map.get(item)));
    }

    public DataBinder<T> bindTextRes(@IdRes int id, ToStringRes<T> map) {
        return bind(p -> (TextView) p.get(id), (item, view) -> view.setText(map.get(item)));
    }

    public DataBinder<T> bindImageRes(@IdRes int id, ToDrawableRes<T> map) {
        return bind(p -> (ImageView) p.get(id), (item, view) -> view.setImageResource(map.get(item)));
    }

    public DataBinder<T> bindImage(@IdRes int id, MapValue<T, String> map) {
        return bind(p -> (ImageView) p.get(id), (item, view) -> loadImage.bind(map.get(item), view));
    }

    public DataBinder<T> bindBackgroundRes(@IdRes int id, ToDrawableRes<T> map) {
        return bind(p -> p.get(id), (item, view) -> view.setBackgroundResource(map.get(item)));
    }

    public DataBinder<T> bindBackground(@IdRes int id, ToDrawable<T> map) {
        return bind(p -> p.get(id), (item, view) -> {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                view.setBackground(map.get(item));
            } else {
                view.setBackgroundDrawable(map.get(item));
            }
        });
    }

    public DataBinder<T> bindVisibility(@IdRes int id, MapValue<T, Boolean> map) {
        return bind(p -> p.get(id), (item, view) -> Ui.visible(view, map.get(item)));
    }

    public DataBinder<T> bindOnClickListener(@IdRes int id, MapValue<T, View.OnClickListener> map) {
        return bind(p -> p.get(id), (item, view) -> view.setOnClickListener(map.get(item)));
    }

    @Override
    public synchronized void onBindViewHolder(BindViewHolder holder, T item, int position) {
        for (IBind<T, BindViewHolder> bind : tasks) {
            bind.onBindViewHolder(holder, item, position);
        }
    }

    public DataBinder setLoadImage(ToView<String, ImageView> loadImage) {
        this.loadImage = loadImage;
        return this;
    }
}

File: Integration/src/main/java/com/inverce/mod/v2/integrations/recycler/DataBinder.kt
package com.inverce.mod.v2.integrations.recycler

import android.graphics.drawable.Drawable
import android.os.Build
import android.support.annotation.IdRes
import android.text.TextWatcher
import android.view.View
import android.widget.EditText
import android.widget.ImageView
import android.widget.TextView
import com.inverce.mod.integrations.R
import com.inverce.mod.v2.core.functional.IMapper
import com.inverce.mod.v2.core.utils.visible
import com.inverce.mod.v2.integrations.view.TextWatcherAdapter
import java.util.*

@Suppress("unused")
open class DataBinder<T> : MultiBind<T, BindViewHolder> {
    public var loadImage: (String, ImageView) -> Unit = { _, _ -> throw IllegalStateException("Image processor not specified") }

    protected var tasks: MutableList<MultiBind<T, BindViewHolder>> = ArrayList()

    fun bind(bind: MultiBind<T, BindViewHolder>) = apply {
        tasks.add(bind)
    }

    inline fun <V : View> bind(crossinline view: IMapper<BindViewHolder, V>, crossinline binder: MultiBind<T, V>) = bind { holder, item, position ->
        binder(view(holder), item, position)
    }

    inline fun <V : View> bind(@IdRes res: Int, crossinline binder: MultiBind<T, V>) = bind { holder, item, position ->
        val v: V = holder[res] ?: throw IllegalStateException("Resource $res not found")
        binder(v, item, position)
    }

    fun bindText(@IdRes id: Int, map: IMapper<T, String>) = bind<TextView>(id) { view, item, _ ->
        view.text = map(item)
    }

    fun bindTextRes(@IdRes id: Int, map: IMapper<T, Int>) = bind<TextView>(id) { view, item, _ ->
        view.setText(map(item))
    }

    fun bindImageRes(@IdRes id: Int, map: IMapper<T, Int>) = bind<ImageView>(id) { view, item, _ ->
        view.setImageResource(map(item))
    }

    fun bindImage(@IdRes id: Int, map: IMapper<T, String>) = bind<ImageView>(id) { view, item, _ ->
        loadImage(map(item), view)
    }

    fun bindBackgroundRes(@IdRes id: Int, map: IMapper<T, Int>) = bind<View>(id) { view, item, _ ->
        view.setBackgroundResource(map(item))
    }

    fun bindBackground(@IdRes id: Int, map: IMapper<T, Drawable>) = bind<View>(id) { view, item, _ ->
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
            view.background = map(item)
        } else {
            @Suppress("DEPRECATION")
            view.setBackgroundDrawable(map(item))
        }
    }

    fun bindVisibility(@IdRes id: Int, map: IMapper<T, Boolean>, gone: Boolean = true) = bind<View>(id) { view, item, _ ->
        view.visible(map(item), gone)
    }

    fun bindOnClickListener(@IdRes id: Int, map: IMapper<T, View.OnClickListener>) = bind<View>(id) { view, item, _ ->
        view.setOnClickListener(map(item))
    }

    fun bindTextChange(id: Int, onChanged: (item: T, text: String) -> Unit) {
        bind<EditText>(id) { txt, it, _ ->
            val inTag = txt.getTag(R.id.im_data_binder_tag_1)
            (inTag as? TextWatcher)?.apply {
                txt.removeTextChangedListener(this)
            }
            val newWatcher = object : TextWatcherAdapter() {
                override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                    onChanged(it, s.toString())
                }
            }
            txt.setTag(R.id.im_data_binder_tag_1, newWatcher)
            txt.addTextChangedListener(newWatcher)
        }
    }

    @Synchronized
    override fun invoke(holder: BindViewHolder, item: T, position: Int) {
        for (bind in tasks) {
            bind(holder, item, position)
        }
    }
}